File: backend/app/schemas/application.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, ConfigDict, Field

from app.models.application import ApplicationStatus


class ApplicationCreate(BaseModel):
    """
    Schema for a student submitting an application to a job posting.
    student_id is NEVER accepted from the client; it is securely derived from current_user.id.
    status is automatically set to 'applied' by the server.
    """
    cover_message: Optional[str] = Field(
        None, max_length=2000, description="Optional cover note or candidate message"
    )


class ApplicationUpdate(BaseModel):
    """
    Schema for a recruiter updating an application's lifecycle status.
    student_id and job_posting_id are immutable and cannot be altered.
    """
    status: ApplicationStatus = Field(
        ..., description="Updated application status (applied, reviewing, shortlisted, rejected, accepted)"
    )


class ApplicationBulkStatusUpdate(BaseModel):
    """
    Schema for a recruiter updating multiple applications in a single atomic batch action.
    """
    application_ids: List[int] = Field(
        ...,
        min_length=1,
        max_length=100,
        description="List of target application IDs to transition (1 to 100 items)",
    )
    status: ApplicationStatus = Field(
        ..., description="Target application status (applied, reviewing, shortlisted, rejected, accepted)"
    )


class ApplicationResponse(BaseModel):
    """
    Safe public response schema for Application.
    Excludes sensitive user credentials.
    """
    id: int
    job_posting_id: int
    student_id: int
    cover_message: Optional[str] = None
    status: ApplicationStatus
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ApplicationBulkStatusResponse(BaseModel):
    """
    Response schema for bulk application status update.
    """
    updated_count: int = Field(..., description="Number of applications updated")
    status: ApplicationStatus = Field(..., description="The new status applied")
    items: List[ApplicationResponse] = Field(..., description="List of updated application objects")

File: frontend/src/types/application.ts
/**
 * Canonical Application Type Definitions (Phase F-06)
 * Strictly aligned with backend schemas in app/schemas/application.py
 * and models in app/models/application.py.
 */

export type ApplicationStatus =
  | 'applied'
  | 'reviewing'
  | 'shortlisted'
  | 'rejected'
  | 'accepted';

export interface Application {
  id: number;
  job_posting_id: number;
  student_id: number;
  cover_message: string | null;
  status: ApplicationStatus;
  created_at: string;
  updated_at: string;
}

export interface ApplicationCreate {
  cover_message?: string;
}

export interface ApplicationUpdate {
  status: ApplicationStatus;
}

export interface ApplicationBulkStatusUpdate {
  application_ids: number[];
  status: ApplicationStatus;
}

export interface ApplicationBulkStatusResponse {
  updated_count: number;
  status: ApplicationStatus;
  items: Application[];
}

/**
 * Filter status options for client-side filtering
 */
export type ApplicationFilterStatus = 'all' | ApplicationStatus;
